itest/rust/src/array_test.rs

```diff
@@ -77,34 +80,56 @@ fn array_into_iterator() {
     assert_eq!(iter.next(), None);
 }
 
+#[itest]
+fn array_hash() {
+    let array = Array::from(&[1, 2]);
+    // Just testing that it converts successfully from i64 to u32.
+    array.hash();
+}
+
 #[itest]
 fn array_clone() {
+    let mut array = Array::from(&[1, 2]);
+    let clone = array.clone();
+    array.set(0, 3.to_variant());
+    assert_eq!(clone.get(0), 3.to_variant());
+}
+
+#[itest]
+fn array_duplicate_shallow() {
     let subarray = Array::from(&[2, 3]);
     let array = Array::from(&[1.to_variant(), subarray.to_variant()]);
-    #[allow(clippy::redundant_clone)]
-    let clone = array.clone();
-    Array::try_from_variant(&clone.get(1))
+    let duplicate = array.duplicate_shallow();
+    Array::try_from_variant(&duplicate.get(1))
         .unwrap()
         .set(0, 4.to_variant());
     assert_eq!(subarray.get(0), 4.to_variant());
 }
 
 #[itest]
-fn array_hash() {
-    let array = Array::from(&[1, 2]);
-    // Just testing that it converts successfully from i64 to u32.
-    array.hash();
+fn array_duplicate_deep() {
+    let subarray = Array::from(&[2, 3]);
+    let array = Array::from(&[1.to_variant(), subarray.to_variant()]);
+    let duplicate = array.duplicate_deep();
+    Array::try_from_variant(&duplicate.get(1))
+        .unwrap()
+        .set(0, 4.to_variant());
+    assert_eq!(subarray.get(0), 2.to_variant());
 }
 
-// TODO: enable once the implementation no longer segfaults
-// #[itest]
-// fn array_duplicate_deep() {
-//     let subarray = Array::from(&[2, 3]);
-//     let array = Array::from(&[1.to_variant(), subarray.to_variant()]);
-//     let mut clone = array.duplicate_deep();
-//     Array::try_from_variant(clone.get(1)).unwrap().set(0, 4.to_variant());
-//     assert_eq!(subarray.get(0), 3.to_variant());
-// }
+#[itest]
+fn array_slice_shallow() {
+    let array = Array::from(&[0, 1, 2, 3, 4, 5]);
+    let slice = array.slice_shallow(5, 1, Some(-2));
+    assert_eq!(slice.try_to_vec::<i64>().unwrap(), vec![5, 3]);
+}
+
+#[itest]
+fn array_slice_deep() {
+    let array = Array::from(&[0, 1, 2, 3, 4, 5]);
+    let slice = array.slice_deep(5, 1, Some(-2));
+    assert_eq!(slice.try_to_vec::<i64>().unwrap(), vec![5, 3]);
+}
```

Comment:
This is the same test as above, which is why it also passes with a shallow implementation 😉

Maybe something like this:
```rs
#[itest]
fn array_slice_deep() {
    let sub_array = Array::from(&[7]);

    let mut array = Array::from(&[0, 1, 2, 3, 4]);
    array.push(sub.to_variant());

    let slice = array.slice_deep(5, 1, Some(-2));
    sub_array.clear(); // modify old array

    let first = slice.get(0).to::<Array>();
    let second = slice.get(1).to::<i64>();
    assert_eq!(first.len(), 1); // should have unmodified array
    assert_eq!(second, 3);
}
```

In fact, the counterpart of this could be used in `array_slice_shallow`.

Reply:
Well, the tests weren't _wrong_ ;) Copied part of the `duplicate_shallow` and `duplicate_deep` tests now.